### PR TITLE
(4.x.x) New implementation of the Transaction Manager

### PR DESCRIPTION
**Pull request by adamretter**
Closes https://github.com/eXist-db/exist/issues/2304

1. Fixes the potential concurrency safety issues in the current transaction manager.
2. New design is non-blocking lock-free.
3. New design does not need to acquire a broker from the Broker Pool for every transaction operation which improves throughput and reduces contention on the Broker Pool.
4. New design allows concurrent  begin, commit, abort operations reducing contention on the Transaction Manager and improving concurrency.

I believe that the design is correct and I have tried to explain it in the Javadoc and with liberal commenting of the code. I recommend that the core developers carefully review my use of Atomics in `TransactionManager` to make sure I have not accidentally introduced new race conditions.

Some testing in a heavily concurrent system would be good too. @wolfgangmm perhaps you could use your concurrent JMeter tests?

### CODE REVIEW

**Comment by duncdrum** [comment]
@adamretter since wolfgangs tests take quite some time, you might want to address codacy `unused import`

**Comment by adamretter** [comment]
@wolfgangmm just wondering if you have seen this?

**Comment by adamretter** [comment]
@wolfgangmm It's been 1 month, any chance you could take a look please?

File: src/org/exist/storage/txn/TransactionManager.java [hunk 1]
```diff
@@ -1,82 +1,131 @@
...
- * This is the central entry point to the transaction management service.
- * 
- * There's only one TransactionManager per database instance that can be
- * retrieved via {@link BrokerPool#getTransactionManager()}. TransactionManager
- * provides methods to create, commit and rollback a transaction.
+ * The Transaction Manager provides methods to begin, commit, and abort
+ * transactions.
+ *
+ * This implementation of the transaction manager is non-blocking lock-free.
+ * It makes use of several CAS variables to ensure thread-safe concurrent
```
Comment (dizzzz) [inline src/org/exist/storage/txn/TransactionManager.java]: CAS ?

File: src/org/exist/storage/txn/TransactionManager.java [hunk 2]
```diff
@@ -1,82 +1,131 @@
...
+     */
+    private final AtomicInteger state = new AtomicInteger(STATE_IDLE);
+
+    /**
+     * Id of the thread which is executing system tasks when
+     * the {@link #state} == {@link #STATE_SYSTEM}. This
+     * is used for reentrancy when system tasks need to
+     * make transactional operations.
+     */
+    private final AtomicLong systemThreadId = new AtomicLong(-1);
```
Comment (dizzzz) [inline src/org/exist/storage/txn/TransactionManager.java]: -1 should be STATE_SYSTEM ?

File: src/org/exist/storage/txn/TransactionManager.java [hunk 3]
```diff
@@ -1,82 +1,131 @@
...
+
+    /**
+     * State for when the Transaction Manager has executing system tasks.
+     */
+    private static final int STATE_SYSTEM = -1;
+
+    /**
+     * State for when the Transaction Manager is idle, i.e. no active transactions.
+     */
+    private static final int STATE_IDLE = 0;
```
Comment (dizzzz) [inline src/org/exist/storage/txn/TransactionManager.java]: could be initialised as `new AtomicLong(0)` so you could potentially perform a faster object identity comparison in certain conditions ? (in stead of value comparison) #notsure :-)

File: src/org/exist/storage/txn/TransactionManager.java [hunk 4]
```diff
@@ -85,63 +134,164 @@ public TransactionManager(final BrokerPool pool, final Optional<JournalManager>
...
     }
     
     /**
      * Commit a transaction.
      * 
-     * @param txn
-     * @throws TransactionException
+     * @param txn the transaction to commit.
+     *
+     * @throws TransactionException if the transaction could not be committed.
```
Comment (dizzzz) [inline src/org/exist/storage/txn/TransactionManager.java]: should `@thows NullPointerException` be added as well?

File: src/org/exist/storage/txn/TransactionManager.java (hunk 1 above)
Comment (adamretter) [reply src/org/exist/storage/txn/TransactionManager.java]: Compare and Swap - e.g. the atomics in Java.

File: src/org/exist/storage/txn/TransactionManager.java (hunk 3 above)
Comment (adamretter) [reply src/org/exist/storage/txn/TransactionManager.java]: @dizzzz I am not sure I follow your reasoning here. Do you want a call to discuss?

File: src/org/exist/storage/txn/TransactionManager.java (hunk 2 above)
Comment (adamretter) [reply src/org/exist/storage/txn/TransactionManager.java]: No. This is the ID of the thread which entered the system state, not the state indicator. -1 just means undefined, as it is an invalid thread id.

File: src/org/exist/storage/txn/TransactionManager.java (hunk 4 above)
Comment (adamretter) [reply src/org/exist/storage/txn/TransactionManager.java]: I don't think that is necessary. Almost every Java method is capable of throwing an NPE.

**Comment by wolfgangmm** [comment]
Using rather simple jmeter tests, I'm able to trigger deadlocks. See stack trace below. I'm trying to see if the issue is reproducible and send the jmeter test if it is.

[jstack.txt](https://github.com/eXist-db/exist/files/2732724/jstack.txt)

**Comment by wolfgangmm** [comment]
The following jmeter test reproducibly triggers the issue for me. Because you need at least one longer running request to encounter the issue, the tests were created against one document in the current version of TEI Publisher (install from public repo). One test generates an epub and will thus lock the document for a while. It completes, but the other waiting threads will never wake up.

Testing against develop-4.x.x all threads complete reliably.

[status.jmx.txt](https://github.com/eXist-db/exist/files/2732939/status.jmx.txt)

**Comment by adamretter** [comment]
@wolfgangmm It's not a deadlock, there are no locks in the new Transaction Manager!

It's a broker starvation issue. I think you have too many incoming request threads for the amount of brokers that are available.

The stack trace shows many threads waiting for a system task to complete, they do that by spinning around here:

```
java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.exist.storage.txn.TransactionManager.beginTransaction(TransactionManager.java:170)
	at org.exist.xmldb.AbstractLocal.withDb(AbstractLocal.java:194)
```

Then there is a single thread, which is executing a system task:

```
"db.exist.scheduler.quartz-worker-1" #25 prio=5 os_prio=31 tid=0x00007fe9c63b3000 nid=0x11603 in Object.wait() [0x000070000b914000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at org.exist.storage.BrokerPool.get(BrokerPool.java:1204)
	- locked <0x0000000740ca6798> (a org.exist.storage.BrokerPool)
	at org.exist.storage.SystemTaskManager.processTasks(SystemTaskManager.java:64)
	- locked <0x0000000740db43a8> (a java.util.ArrayDeque)
	at org.exist.storage.txn.TransactionManager.processSystemTasks(TransactionManager.java:536)
	at org.exist.storage.txn.TransactionManager.triggerSystemTask(TransactionManager.java:521)
```

The system task cannot proceed because there are no brokers available for it to use.

Before this PR each of the waiting threads would (as they still do in this PR) take a broker in `AbstractLocal.withDb` and then block when they enter `TransactionManager#beginTransaction` waiting on an exclusive lock in the transaction manager. The difference is that before this PR, a DBBroker would be taken in `TransactionManager#triggerSystemTasks` before the exclusive lock was taken in the transaction manager.

There are two solutions to solving the problem in the PR:

1. `TransactionManager#processSystemTasks` must take a broker before switching into the SYSTEM state, so that the system task cannot be starved of brokers. This is the simplest but a bit yucky, as the transaction manager shouldn't have to be concerned with this. A small change to the API where the broker is passed to `SystemTaskManager#processTasks` could make this slightly less yucky through making the broker contract explicit.

2. We reserve a broker in the broker pool which is only used for system tasks, and use that broker. This requires a bit more work but is probably a cleaner approach.

Right now, I am short on time and so could implement (1).

**Comment by wolfgangmm** [comment]
Ok, so this explains why I need 20 threads to bring the system to a halt and why the same works reliably on develop-4.x.x. I'll investigate and see if I could work on option 2).

**Comment by adamretter** [comment]
@wolfgangmm I think it's better just to do (1) now and focus on other things. I will send a commit for (1) I have it tested and working here.

**Comment by adamretter** [comment]
@wolfgangmm okay it is pushed. Could you test again please?

If you're then happy with it, I will cherry-pick the additional commit to the other branches...

**Comment by wolfgangmm** [comment]
I tested again and now everything runs fine, thanks!

**Comment by dizzzz** [comment]
buildfailure can be ignored.

**Pull request merged**